Use prettier to format source files

mesaugat opened the pull request:
* Added `prettier` and `tslint-config-prettier` as dev dependencies
* Added a npm script to run prettier from the CLI: `npm run prettier`
* Formats `.js`, `.ts`, `.tsx`, `.json` and `.md` files
* Added `.prettierrc` file with a basic configuration
* Included a `.prettierignore` for ignoring files that don't require prettier
* `tslint-config-prettier` has been added to `tslint.json`. It helps in the removal of conflicts between `tslint` rules and `prettier`
* Another npm script has been added to check for conflicting rules `npm run tslint:check`. The following rules were detected as conflicts by `tslint-config-prettier` and have been removed from `tslint.json`
```
  max-line-length
  no-irregular-whitespace
  no-unnecessary-semicolons
  number-literal-format
  object-literal-key-quotes
  quotemark
  space-within-parens
  trailing-comma
  whitespace
```
* `tslint:check` has been added to the `test` script
* `recommended_ruleset.js` is now generated using single quotes. Also, to avoid conflicts with prettier, `recommended_ruleset.js` has been added to `.prettierignore`
* All files have been formatted using prettier
* Updated tests after running prettier
* Added `husky` and `lint-staged` to run prettier as a pre-commit hook

Fixes #558

mesaugat commented:
I forgot to run the tests, it's failing. 😞

IllusionMH commented:
Fitrst of all - thanks for efforts and PR, unified formatting is great.

This project used single quotes that were enforced with TSLint and as I can see - there are many files with only quotes changes.
IMHO it makes sense to add `singleQuote: true` to `.prettierrc` file.

Other things to consider:
- add `prettier` to that end of `test` script.
  There are several files that are created during `npm test` that will have conflicting formatting (`rule-metadata.json`, `recommended_ruleset.js`) and would be great to have them formatted accordingly.
  Do you know how long `npm run prettier` takes?

- Update `build-tasks/templates/rule.template.js` to match formatting or run prettier after it (I can help with passing arguments to scripts within `npm-run-all`).

- Update build tasks to output `JSON` with `4` spaces instead of `2` (not necessary if `prettier` script will be added to `test` and `create-rule` but nice to have)

IllusionMH commented on build-tasks/templates/rule.template.js:
```diff
@@ -3,9 +3,7 @@ module.exports = ({ ruleName, walkerName }) =>
 import {ExtendedMetadata} from './utils/ExtendedMetadata';
-// use (and contribute to) AstUtils for common AST functions // TODO: delete comment
 import {AstUtils} from './utils/AstUtils';
-// use Utils instead of Underscore functions // TODO: delete comment
```
Looks like it is comment for developer who is creating new rule. Not sure if they should be removed

mesaugat replied:
Ah okay, I thought so. My bad.

mesaugat commented:
I'll fix the tests too.

IllusionMH commented on src/exportNameRule.ts:
```diff
@@ -162,7 +157,7 @@ export class ExportNameWalker extends Lint.RuleWalker {
     private validateExport(exportedName: string, node: ts.Node): void {
         const flags = Rule.getIgnoreCase(this.getOptions()) ? 'i' : '';
-        const regex : RegExp = new RegExp(exportedName + '\..*', flags); // filename must be exported name plus any extension
+        const regex: RegExp = new RegExp(exportedName + '..*', flags); // filename must be exported name plus any extension
```
interesting change, may be there was an error and it should be `\\..`

mesaugat replied:
Should I make it `\\`? This change was made by prettier.

JoshuaKGoldberg commented on package.json:
```diff
@@ -49,11 +49,13 @@
     "test:rules": "tslint -r dist/src --test \"tests/**\"",
     "test": "npm-run-all clean copy:json tsc:src test:* lint:* validate:* copy:package copy:meta generate:*",
     "tsc:src": "tsc",
+    "tslint:check": "tslint-config-prettier-check ./tslint.json",
```
Let's add this to the `test` command so we remember to run it.

JoshuaKGoldberg commented on package.json:
```diff
@@ -73,9 +75,11 @@
     "glob": "^7.1.3",
     "mocha": "5.2.0",
     "npm-run-all": "^4.1.3",
+    "prettier": "^1.14.3",
```
Ooh, I've been bitten by version ranges in auto-formatters before. Please remove the `^`.

mesaugat replied:
Sure.

JoshuaKGoldberg commented on tslint.json:
```diff
@@ -1,221 +1,191 @@
+        "react-tsx-curly-spacing": false,
+        "valid-typeof": false
+    }
 }
```
These rules can be completely removed from `tslint.json` because they're mentioned in `tslint-config-prettier`, right?

max-line-length
object-literal-key-quotes
quotemark
trailing-comma
whitespace
align
arrow-parens
arrow-return-shorthand (?)
space-before-function-name
space-before-function-paren

mesaugat replied:
Yes, it can be removed.

mesaugat commented:
The tests are passing. 🎉 

@IllusionMH @JoshuaKGoldberg Should I add `prettier` to the end of `test` script? It takes a good `10-15` seconds to run prettier on all files.

Maybe we can use a [pre-commit hook](https://prettier.io/docs/en/precommit.html) to run prettier.

JoshuaKGoldberg commented:
Yes please, a pre-commit hook would be best. Husky is pretty great.

JoshuaKGoldberg commented on .vscode/extensions.json:
```diff
@@ -1,9 +1,9 @@
+    "recommendations": [
+        // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+        "EditorConfig.EditorConfig",
+        "eg2.tslint"
```
Ah, just noticed - should we also recommend the Prettier VS Code extension?

JoshuaKGoldberg replied:
Ah, just noticed - should we also recommend the Prettier VS Code extension?

IllusionMH commented:
Also concatenation can be removed in next lines, since now that are on one line anyway (probably in follow up PR)
```
build-tasks/generate-sdl-report.js:13
src/chaiVagueErrorsRule.ts:10
src/chaiVagueErrorsRule.ts:12
src/insecureRandomRule.ts:7
src/insecureRandomRule.ts:9
src/jqueryDeferredMustCompleteRule.ts:26
src/missingJsdocRule.ts:28
src/noDuplicateCaseRule.ts:30
src/reactA11yRoleSupportsAriaPropsRule.ts:43
src/reactA11yRoleSupportsAriaPropsRule.ts:12
```

mesaugat commented on src/utils/attributes/README.md:
```diff
@@ -1,14 +1,14 @@
-  This file is composed of aria-* attributes as well as their types or limited values if any.
-  Each aria-* attribute has a required type.
-  Some attributes has a group of supported values, they MUST only use one of these values.
+   This file is composed of aria-_ attributes as well as their types or limited values if any.
```
Resolved.

mesaugat commented:
@IllusionMH I've removed the concatenations too. These were just some aesthetic changes thus pushed it here.

IllusionMH commented:
> I've removed the concatenations too.
Great!

I've checked on Windows in VS Code. Both pre-commit and `"editor.formatOnSave": true,` works great with Prettier.

@mesaugat Thanks! There are no more notes from my side 👍

JoshuaKGoldberg commented:
Ahh, merge conflicts. I'll try to get to these tonight if I have permissions for the branch.

IllusionMH commented:
👍 That was fast conflict resolving 😮

JoshuaKGoldberg commented:
Tested this out locally and it works like a charm. Thanks so much @mesaugat for sending this in and working on it, and @IllusionMH for the great reviews! 🙌

mesaugat commented:
@IllusionMH @JoshuaKGoldberg Thank you guys! 🥇

Also, I forgot to mention but I had a hard time making changes to the tests when doing this. Somehow, the tests showed an old snapshot of `expected/actual` even though I changed some of the tests.

I've not looked into the test utils but maybe the test results were being cached or something. These [tests](https://github.com/Microsoft/tslint-microsoft-contrib/blob/master/src/tests/NoReservedKeywordsTests.ts) used to fail even though I changed it to make is green.
